File: backend/pkg/board/board.go
package board

import (
	"errors"
)

const (
	RED          = 1
	BLACK        = 2
	BOARD_WIDTH  = 7
	BOARD_HEIGHT = 6
)

var runeIntMap = map[rune]int{
	'0': 0,
	'1': 1,
	'2': 2,
	'3': 3,
	'4': 4,
	'5': 5,
	'6': 6,
}
var intStringMap = map[int]string{
	0: "0",
	1: "1",
	2: "2",
	3: "3",
	4: "4",
	5: "5",
	6: "6",
}

type Board struct {
	Data []int
}

func BoardFromString(moves string) (board Board) {
	for _, runeValue := range moves {
		board.Data = append(board.Data, runeIntMap[runeValue])
	}
	return
}

func (b *Board) String() (out string) {
	for _, column := range b.Data {
		out += intStringMap[column]
	}
	return
}

func (b *Board) PlayMove(column int) error {
	if len(b.Data) == 6*7 {
		return errors.New("Board full")
	}
	var count int
	for _, move := range b.Data {
		if move == column {
			count += 1
		}
	}
	if count == 6 {
		return errors.New("Column full")
	}
	b.Data = append(b.Data, column)
	return nil
}

func (b *Board) WhoIsNext() int {
	if len(b.Data)%2 == 1 {
		return BLACK
	}
	return RED
}

func (b *Board) ClockShouldBeStarted() bool {
	return len(b.Data) > 1
}

func (b *Board) asMatrix() (board [7][6]int) {
	var counters [7]int
	for index, column := range b.Data {
		board[column][counters[column]] = (index % 2) + 1
		counters[column] += 1
	}
	return
}

func (b *Board) IsWon() int {
	board := b.asMatrix()
	// horizontalCheck
	for j := 0; j < BOARD_HEIGHT-3; j++ {
		for i := 0; i < BOARD_WIDTH; i++ {
			if (board[i][j] == RED || board[i][j] == BLACK) &&
				board[i][j] == board[i][j+1] &&
				board[i][j] == board[i][j+2] &&
				board[i][j] == board[i][j+3] {
				return board[i][j]
			}
		}
	}
	// verticalCheck
	for i := 0; i < BOARD_WIDTH-3; i++ {
		for j := 0; j < BOARD_HEIGHT; j++ {
			if (board[i][j] == RED || board[i][j] == BLACK) &&
				board[i][j] == board[i+1][j] &&
				board[i][j] == board[i+2][j] &&
				board[i][j] == board[i+3][j] {
				return board[i][j]
			}
		}
	}
	// ascendingDiagonalCheck
	for i := 3; i < BOARD_WIDTH; i++ {
		for j := 0; j < BOARD_HEIGHT-3; j++ {
			if (board[i][j] == RED || board[i][j] == BLACK) &&
				board[i][j] == board[i-1][j+1] &&
				board[i][j] == board[i-2][j+2] &&
				board[i][j] == board[i-3][j+3] {
				return board[i][j]
			}
		}
	}
	// descendingDiagonalCheck
	for i := 3; i < BOARD_WIDTH; i++ {
		for j := 3; j < BOARD_HEIGHT; j++ {
			if (board[i][j] == RED || board[i][j] == BLACK) &&
				board[i][j] == board[i-1][j-1] &&
				board[i][j] == board[i-2][j-2] &&
				board[i][j] == board[i-3][j-3] {
				return board[i][j]
			}
		}
	}
	return 0
}

File: mobile-app/src/board.js
/**
 * @prettier
 * @flow
 */

const RED = 1;
const BLACK = 2;
const BOARD_WIDTH = 7;
const BOARD_HEIGHT = 6;

const stringIntMap = {
    "0": 0,
    "1": 1,
    "2": 2,
    "3": 3,
    "4": 4,
    "5": 5,
    "6": 6,
};

// type Board struct {
//     Data []int
// }
class Board {
    data: Array<number>;
    constructor(data: ?Array<number>) {
        this.data = data || [];
    }

    playMove(column: number): ?Error {
        if (this.data.length == BOARD_WIDTH * BOARD_HEIGHT) {
            return new Error("Board full");
        }
        let count = 0;
        for (var i = 0; i < this.data.length; i++) {
            if (this.data[i] == column) {
                count++;
            }
        }
        if (count == BOARD_HEIGHT) {
            return new Error("Column full");
        }
        this.data.push(column);
    }
    whoIsNext(): number {
        if (this.data.length % 2 == 1) {
            return BLACK;
        }
        return RED;
    }

    asMatrix(): Array<Array<number>> {
        let counters = new Array(BOARD_WIDTH).fill(0);
        let board = new Array(BOARD_WIDTH).fill(
            new Array(BOARD_HEIGHT).fill(0),
        );
        for (var i = 0; i < this.data.length; i++) {
            const column = this.data[i];
            board[column][counters[column]] = i % 2 + 1;
            counters[column] += 1;
        }
        return board;
    }
    isWon(): number {
        const board = this.asMatrix();
        // horizontalCheck
        for (let j = 0; j < BOARD_HEIGHT - 3; j++) {
            for (let i = 0; i < BOARD_WIDTH; i++) {
                if (
                    (board[i][j] == RED || board[i][j] == BLACK) &&
                    board[i][j] == board[i][j + 1] &&
                    board[i][j] == board[i][j + 2] &&
                    board[i][j] == board[i][j + 3]
                ) {
                    return board[i][j];
                }
            }
        }
        // verticalCheck
        for (let i = 0; i < BOARD_WIDTH - 3; i++) {
            for (let j = 0; j < BOARD_HEIGHT; j++) {
                if (
                    (board[i][j] == RED || board[i][j] == BLACK) &&
                    board[i][j] == board[i + 1][j] &&
                    board[i][j] == board[i + 2][j] &&
                    board[i][j] == board[i + 3][j]
                ) {
                    return board[i][j];
                }
            }
        }
        // ascendingDiagonalCheck
        for (let i = 3; i < BOARD_WIDTH; i++) {
            for (let j = 0; j < BOARD_HEIGHT - 3; j++) {
                if (
                    (board[i][j] == RED || board[i][j] == BLACK) &&
                    board[i][j] == board[i - 1][j + 1] &&
                    board[i][j] == board[i - 2][j + 2] &&
                    board[i][j] == board[i - 3][j + 3]
                ) {
                    return board[i][j];
                }
            }
        }
        // descendingDiagonalCheck
        for (let i = 3; i < BOARD_WIDTH; i++) {
            for (let j = 3; j < BOARD_HEIGHT; j++) {
                if (
                    (board[i][j] == RED || board[i][j] == BLACK) &&
                    board[i][j] == board[i - 1][j - 1] &&
                    board[i][j] == board[i - 2][j - 2] &&
                    board[i][j] == board[i - 3][j - 3]
                ) {
                    return board[i][j];
                }
            }
        }
        return 0;
    }
}

const BoardFromString = (moves: string): Board => {
    let board = new Board();
    for (var i = 0; i < moves.length; i++) {
        board.data.push(stringIntMap[moves]);
    }
    return board;
};

module.exports = {
    Board: Board,
    BoardFromString: BoardFromString,
};
